### PR TITLE
Update to support ESLint 9

### PR DESCRIPTION
**Pull request by christopher-buss**
This updates the plugin to allow it to be used with ESLint 9, while not breaking functionality for existing users. Originally this PR was meant to support ESLint 9 directly, but after consideration, typescript-eslint currently still only officially supports ESLint 8, therefore there is no guarentees about the stability of upgrading currently. Once typescript-eslint officially supports ESLint 9, a new PR can be made to bump our dependencies then, and officially support ESLint 9.

The now deprecated `@typescript-eslint/experimental-utils` has been changed to @typescript-eslint/utils.

### CODE REVIEW

File: src/rules/lua-truthiness.ts [hunk 1]
```diff
@@ -9,7 +10,7 @@ export const luaTruthiness = makeRule<[], "falsyStringNumberCheck">({
 		type: "problem",
 		docs: {
 			description: "Warns against falsy strings and numbers",
-			recommended: false,
+			recommended: "recommended",
```
Comment (Dionysusnu) [inline src/rules/lua-truthiness.ts:12]: This looks like a behaviour change?

File: src/rules/lua-truthiness.ts [hunk 2]
```diff
@@ -1,4 +1,5 @@
-import { AST_NODE_TYPES, TSESTree } from "@typescript-eslint/experimental-utils";
+
```
Comment (Dionysusnu) [inline src/rules/lua-truthiness.ts]: Empty line at the start of the file
(Does prettier not fix this? It looks like it is configured in .eslintrc, although I note that there's no CI flow for either of them in this repo)

File: .gitignore [hunk 3]
```diff
@@ -1,3 +1,4 @@
 node_modules
 out
 .vscode/settings.json
+*.tgz
```
Comment (Dionysusnu) [inline .gitignore:4]: Missing a newline at the end of the file.

I'm guessing this change is for `npm pack` testing?

File: package.json [hunk 4]
```diff
@@ -9,20 +9,20 @@
...
+		"eslint": "set ESLINT_USE_FLAT_CONFIG=false eslint \"src/**/*.ts\" --max-warnings 0 "
 	},
 	"dependencies": {
-		"@types/node": "^20.8.0",
-		"@typescript-eslint/experimental-utils": "^5.62.0",
-		"typescript": "^5.2.2"
+		"@types/node": "^20.12.0",
+		"@typescript-eslint/utils": "^7.8.0",
+		"typescript": "^5.4.0"
 	},
```
Comment (Dionysusnu) [inline package.json:18]: Should this include a peerDependency to eslint 9? I feel like, without that, people will update the plugin without updating eslint itself, and then run into errors. Or are these changes backwards compatible with lower eslint versions?

File: src/rules/misleadingLuatupleChecks.ts [hunk 5]
```diff
@@ -19,11 +19,11 @@ export const misleadingLuatupleChecks = makeRule<[], "bannedLuaTupleCheck">({
 	},
 	defaultOptions: [],
 	create(context) {
-		const service = getParserServices(context);
+		const service = ESLintUtils.getParserServices(context);
 		const checker = service.program.getTypeChecker();
 
 		function checkTruthy(node: TSESTree.Node) {
-			const { aliasSymbol } = getConstrainedType(service, checker, node);
+			const { aliasSymbol } = getConstrainedType(service, checker!, node);
```
Comment (Dionysusnu) [inline src/rules/misleadingLuatupleChecks.ts:26]: Leftover type assertion, is this still necessary?

File: src/rules/noPrivateIdentifier.ts [hunk 6]
```diff
@@ -19,7 +20,7 @@ export const noPrivateIdentifier = makeRule<[], "privateIdentifierViolation">({
 	},
 	defaultOptions: [],
 	create(context) {
-		const service = getParserServices(context);
+		const service = ESLintUtils.getParserServices(context);
```
Comment (Dionysusnu) [inline src/rules/noPrivateIdentifier.ts:23]: Whitespace on the end of the line

**Review by Dionysusnu — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 3 above)
Comment (christopher-buss) [reply .gitignore:4]: Yes, this was for my testing purposes as I wanted to ensure the plugin worked. I can leave it out if preferred.

File: package.json (hunk 4 above)
Comment (christopher-buss) [reply package.json:18]: I'm currently unsure. I'll do a test with the old eslint version to check if this works. These changes should still work with ESLint versions > 8, but I'll do a quick review and get back here.

File: src/rules/lua-truthiness.ts (hunk 2 above)
Comment (christopher-buss) [reply src/rules/lua-truthiness.ts]: I'm unsure what happened here, my latest revision to these changes no longer has this issue.

File: package.json (hunk 4 above)
Comment (christopher-buss) [reply package.json:18]: I have checked, and using the default create roblox-ts, this works with eslint 8.57.0 out the box, but does not work with eslint version, I'll add a peerDependency for eslint 8.56.0. I'm not sure technically what the earlier eslint version supported is, but I tried 8.21.0 (where the flat config files were originally added), and this does not work. The latest versions of the eslint-plugin and parser require ^8.56.0, so it's probably a good enough place to require our users to be.

File: package.json [hunk 7]
```diff
@@ -9,7 +9,7 @@
 		"build": "tsc",
 		"build-watch": "tsc -w",
 		"prepublishOnly": "npm run build",
-		"eslint": "set ESLINT_USE_FLAT_CONFIG=false eslint \"src/**/*.ts\" --max-warnings 0 "
+		"eslint": "cross-env ESLINT_USE_FLAT_CONFIG=false eslint .\\src\\"
```
Comment (Dionysusnu) [inline package.json]: This has removed the `--max-warnings 0`, which is relevant for CI usage

File: package.json [hunk 8]
```diff
@@ -25,6 +25,9 @@
 		"eslint-plugin-prettier": "^5.1.0",
 		"prettier": "^3.2.3"
 	},
+	"peerDependencies": {
+		"eslint": "^8.56.0"
```
Comment (Dionysusnu) [inline package.json:28]: Should this be `>` rather than `^`, so that people _can_ use eslint 9 if they prefer?

File: pnpm-lock.yaml
Comment (Dionysusnu) [inline pnpm-lock.yaml:1]: Unintentionally committed?

File: src/rules/lua-truthiness.ts [hunk 9]
```diff
@@ -9,7 +9,6 @@ export const luaTruthiness = makeRule<[], "falsyStringNumberCheck">({
 		type: "problem",
 		docs: {
 			description: "Warns against falsy strings and numbers",
-			recommended: false,
```
Comment (Dionysusnu) [inline src/rules/lua-truthiness.ts:12]: Is it possible to be explicit in this field, or does the new string option only allow "recommended" or <absent>?

File: src/rules/lua-truthiness.ts [hunk 10]
```diff
@@ -21,11 +20,11 @@ export const luaTruthiness = makeRule<[], "falsyStringNumberCheck">({
 	},
 	defaultOptions: [],
 	create(context) {
-		const service = getParserServices(context);
+		const service = ESLintUtils.getParserServices(context);
 		const checker = service.program.getTypeChecker();
 
 		function checkTruthy(node: TSESTree.Node) {
-			const type = getType(checker, service.esTreeNodeToTSNodeMap.get(node));
+			const type = getType(checker!, service.esTreeNodeToTSNodeMap.get(node));
```
Comment (Dionysusnu) [inline src/rules/lua-truthiness.ts:27]: Leftover type assertion that I missed in my previous review

File: package.json [hunk 11]
```diff
@@ -9,20 +9,23 @@
...
 	"devDependencies": {
-		"@typescript-eslint/eslint-plugin": "^6.7.3",
-		"@typescript-eslint/parser": "^6.7.3",
-		"eslint": "^8.50.0",
-		"eslint-config-prettier": "^9.0.0",
-		"eslint-plugin-prettier": "^5.0.0",
-		"prettier": "^3.0.3"
+		"@typescript-eslint/eslint-plugin": "^7.8.0",
+		"@typescript-eslint/parser": "^7.8.0",
+		"eslint": "8.57.0",
```
Comment (Dionysusnu) [inline package.json:22]: Style: npm uses `^` by default if just a version number is given, but adding it explicitly is easier to read and consistent with the other dependencies

**Review by Dionysusnu — CHANGES_REQUESTED** [review]

File: package.json (hunk 8 above)
Comment (christopher-buss) [reply package.json:28]: I didn't know this was a thing!

File: src/rules/lua-truthiness.ts (hunk 9 above)
Comment (christopher-buss) [reply src/rules/lua-truthiness.ts:12]: Yes. Recommended or absent are the only options here.

**Comment by christopher-buss** [comment]
I've made the releveant changes as required! Let me know if there's anything else.

**Review by Dionysusnu — APPROVED** [review]

**Comment by christopher-buss** [comment]
I've closed this so it doesn't accidentally get merged. This works for me because I'm manually defining each rule, but removing the "recommended" section in the way I have will cause the code that generates the ruleset to not work. TypeScript ESLint utils v8 exports better ways of doing this, but it's currently in beta. I'll either fix this PR up, or wait until the utils is ready and create a proper release.

**Pull request closed without merging**